Adding filters to triggers using k8s LabelSelectors

nachocano opened the pull request:
## Proposed Changes

- Adding filters to triggers using k8s Label Selectors. With this approach, we can reuse all the existing machinery for matching labels.
- As k8s LabelSelectors validate label keys and values, we are bypassing that validation for now. In order to match expressions, we will probably need to write our own "FilterSelectorRequirement". In a subsequent PR we will prototype a quick hack using reflection to set unexported fields in a Requirement struct. 

A trigger example looks like this:

```yaml
apiVersion: eventing.knative.dev/v1alpha1
kind: Trigger
metadata:
  name: trigger-test
spec:
  broker: broker-test
  filter:
    headers:
      Ce-Eventtype: '"com.example.someevent"'
      Ce-Source: '"/mycontext/subcontext"'
  subscriber:
    ref:
      apiVersion: serving.knative.dev/v1alpha1
      kind: Service
      name: message-dumper

```

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
NONE
```

nachocano committed:
Adding filtering using k8s label selectors

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new-model

nachocano committed:
Updating trigger example with filters

nachocano commented:
/cc @Harwayne

nachocano commented:
@Harwayne  I guess the /cc has to be enabled in the repo. Let me know if it makes sense. We can chat more tomorrow.
Thanks!

Harwayne commented:
All the /commands are prow commands as far as I know. I think using @<name> will inform people (depending on their email settings).

Harwayne commented on t2.yaml:
```diff
@@ -3,12 +3,14 @@ kind: Trigger
      kind: Service
      name: message-dumper
-
+  # Our Cloud Event parsing library seems to have a bug and forces to put type and source in double
```
I think this means that that the values below need to be wrapped in double quotes. Have you tried running this to see what passes through?

nachocano replied:
Done, now it works as your example.

Harwayne commented on t2.yaml:
```diff
@@ -3,12 +3,14 @@ kind: Trigger
-
+  # Our Cloud Event parsing library seems to have a bug and forces to put type and source in double
+  # quotes (as it thinks the actual value is `"foo"`, not `foo`).
+  filters:
```
My preference is to put `filters` before `subscriber`.

nachocano replied:
Done

Harwayne commented on pkg/apis/eventing/v1alpha1/trigger_types.go:
```diff
@@ -58,11 +58,10 @@ type TriggerSpec struct {
 
-	Type   string `json:"type,omitempty"`
-	Source string `json:"source,omitempty"`
+	Filters *metav1.LabelSelector `json:"filters,omitempty"`
```
Let's put in another layer of indirection:

```yaml
spec:
  filter:
    headers:
      Ce-Eventtype: com.example.someevent
      ...
```

That gives an easy place to add something like `headerExpressions`, as sibling of `headers`

nachocano replied:
Done. Having headerExpressions will be sligthly more involved than I thought. There is a Requirement struct that doesn't expose its fields. I'm trying to do a quick prototype using reflection to set those values. For now, I just pushed the headers with exact matching. It works as what you previously had of type and source.

Harwayne reviewed (commented):
Looks awesome!

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new-model

nachocano committed:
Updating after review comments. Only doing exact header matching.
In a follow up PR, I plan to include matching of expressions. It turns
our that they are somewhat more involved than expected.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new-model

nachocano commented:
@Harwayne Done the changes. I still have locally my headerExpressions part. I'm doing something ugly with reflection in order to set some unexported fields. Will probably create another PR with that, because you won't like it, and we may need to add our own implementation. I guess this part works as the current one, but with a slightly better definition in the yaml.
